core/libp2p/multi/multiaddress.cpp

```diff
@@ -152,8 +140,55 @@ namespace libp2p::multi {
     return values;
   }
 
+  std::list<Protocol> Multiaddress::getProtocols() const {
+    std::string_view addr{stringified_address_};
+    addr.remove_prefix(1);
+
+    std::list<std::string> tokens;
+
+    boost::algorithm::split(tokens, addr, boost::algorithm::is_any_of("/"));
+
+    std::list<Protocol> protocols;
+    for (auto &token : tokens) {
+      auto p = ProtocolList::get(token);
+      if (p) {
+        protocols.emplace_back(*p);
+      }
+    }
+    return protocols;
+  }
+
+  std::list<std::pair<Protocol, std::string>>
+  Multiaddress::getProtocolsWithValues() const {
+    std::string_view addr{stringified_address_};
+    addr.remove_prefix(1);
+    if (addr.back() == '/')
+      addr.remove_suffix(1);
+
+    std::list<std::string> tokens;
+
+    boost::algorithm::split(tokens, addr, boost::algorithm::is_any_of("/"));
+
+    std::list<std::pair<Protocol, std::string>> pvs;
+    for (auto &token : tokens) {
+      auto p = ProtocolList::get(token);
+      if (p) {
```

Comment:
here you assume that if ProtocolList::get returns value (and not nullptr), then it is protocol name; and in `else` clause, you assume that it is value.

This approach is wrong. You should use queue or list (as queue) and while(!queue.empty()) pop 2 first elements (if can't pop 2 elements - multiaddr is invalid, because there is a key, but no value)

Reply:
It's probably the major problem I encountered. There are multiaddresses that look like "/ip4/1.2.3.4/tcp/443/tls/sni/example.com/http/example.com/index.html". As I understand, there is key "http" with value "example.com/index.html" and here "/tls/sni/example.com" there seem to be something different from /key/value, too.